### PR TITLE
Add checks ADR

### PR DESCRIPTION
**Pull request by fabriziosestito**
Adding some ADR about the checks architectural decisions that we took when re-designing the checks engine

### CODE REVIEW

File: adr/0006-use-cloudevents-to-describe-event-data.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+## Context
+
+We need to define a common way to describe the data of the events that are exchanged between the components of Trento.
+An envelope format is needed to carry the event payload and metadata and to easily serialize the data for the transport layer.
+Ideally, we want to use a format that allows us to open the door for third-party components to easily integrate with Trento.
+
+## Decision
+
+Decided on [CloudEvents](https://cloudevents.io/).
```
Comment (dottorblaster) [inline adr/0006-use-cloudevents-to-describe-event-data.md:17]:
```suggestion
The team chose [CloudEvents](https://cloudevents.io/).
```

**Review by dottorblaster — APPROVED** [review]
LGTM

File: adr/0003-add-facts-gathering-capabilities-to-the-agent.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+We need to implement a way to run the gatherers and collect the facts, namely the "gathering engine".
+Also, advanced users might want to declare their gatherers to be used in the expectations section,
+we can refer to this as "gatherers plugins".
+
+## Decision
+
+The agent should be able to gather facts from the hosts and send them to the server.
+A request is sent from the server to the agent to gather facts, the agent runs the gatherers and sends the facts to the server.
+The communication between the agent and the server is asynchronously done via AMQP.
+If the agent is not able to gather one or more facts, it should send a message to the server with the list of the failed gatherers alongside the successfully gathered facts, in a best-effort fashion.
```
Comment (arbulu89) [inline adr/0003-add-facts-gathering-capabilities-to-the-agent.md:22]: I would add a last sentence saying that this allows the agent to query only once the "source of truth" of specific gatherers. For example: one query once the cibadmin command for each fact gathering process.
This helps to reduce any potential overload in the agents.
In the past, with ansible, without having this "cache" capability, we were running the commands for each particular check.

File: adr/0005-use-protobuf-to-define-and-generate-contracts.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+## Context
+
+We need to define contracts between the components of Trento to communicate with each other.
+We want to use a well-known and widely adopted format to define these contracts.
+We want to use a format that allows us to generate code from the contracts, avoid boilerplate code and reduce the risk of errors.
+
+## Decision
+
+Protocol Buffers is a good candidate for this task, as it is a language-neutral, platform-neutral, extensible mechanism for serializing structured data.
```
Comment (arbulu89) [inline adr/0005-use-protobuf-to-define-and-generate-contracts.md:17]: Could we put a link to protobuf together with the `Protocol Buffers` name?
https://developers.google.com/protocol-buffers/
Or this commonly used `protobuf` term. We are used to this word, and it simplifies questions (like; "Is this protocol buffers "protobuf"")
Yes, I know, we should know that this refers to the `protobuf`

**Review by arbulu89 — APPROVED** [review]
Great @fabriziosestito ,
Really appreciated the effort here

File: adr/0003-add-facts-gathering-capabilities-to-the-agent.md (hunk 2 above)
Comment (fabriziosestito) [reply adr/0003-add-facts-gathering-capabilities-to-the-agent.md:22]: nice i will add it!

File: adr/0005-use-protobuf-to-define-and-generate-contracts.md (hunk 3 above)
Comment (fabriziosestito) [reply adr/0005-use-protobuf-to-define-and-generate-contracts.md:17]: protobuf is the short for Protocol Buffers, they are both used. I can do what Wikipedia does: Protocol Buffers (Protobuf) so we all know what we are referring to

**Review by nelsonkopliku — COMMENTED** [review]
Really good job. Just a couple of comments.

File: adr/0003-add-facts-gathering-capabilities-to-the-agent.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+We need to gather facts from the hosts to be able to use them in the expectations section of the checks.
+A "gatherer" is a piece of code that knows how to extract a specific fact from a host.
+We need to implement a way to run the gatherers and collect the facts, namely the "gathering engine".
+Also, advanced users might want to declare their gatherers to be used in the expectations section,
+we can refer to this as "gatherers plugins".
+
+## Decision
+
+The agent should be able to gather facts from the hosts and send them to the server.
```
[the remark this answers is not in the document]
Comment (fabriziosestito) [reply adr/0003-add-facts-gathering-capabilities-to-the-agent.md:19]: In out language server usually refers to whatever does not run in the agent and it's outside the workload (prometheus, dashboard, runner, etc..). We can be more precise here but I wanted to be intentionally vague to not couple the agent fact capabilities to the checks execution concept.

File: adr/0004-build-a-check-execution-orchestrator.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
+# 4. Build a check execution orchestrator
```
[the remark this answers is not in the document]
Comment (fabriziosestito) [reply adr/0004-build-a-check-execution-orchestrator.md:1]: makes sense

File: adr/0005-use-protobuf-to-define-and-generate-contracts.md (hunk 3 above)
Comment (arbulu89) [reply adr/0005-use-protobuf-to-define-and-generate-contracts.md:17]: Yes, I would do that. We use `protobuf` always, so if somebody does some search or anything in the ADRs folder project

**Pull request merged**